### PR TITLE
[DWARFVerifier] Allow overlapping ranges for ICF-merged functions

### PR DESCRIPTION
**Pull request by alx32**
This patch modifies the DWARF verifier to handle a valid case where two or more functions have identical address ranges due to being merged by ICF (Identical Code Folding). Previously, the verifier would incorrectly report these as errors, but functions merged via ICF (such as when using LLD's --keep-icf-stabs option) can legitimately share the same address range.

A new test case has been added to verify this behavior using YAML-based DWARF data that simulates two DW_TAG_subprogram entries with identical address ranges. The test ensures that the verifier correctly identifies this as a valid case and doesn't emit any errors, while still maintaining the existing verification for truly invalid overlapping ranges in other scenarios. Before this change, the newly added test case would have failed, with `llvm-dwarfdump` marking the overlapping address ranges in the DWARF as an error. 

We also modify the existing test `llvm-dwarfutil/ELF/X86/verify.test` which relies on the existence of the error that we're trying to suppress. Here we slightly change one offset so that the ranges don't perfectly overlap and an error is still generated.

### CODE REVIEW

**Comment by dwblaikie** [comment]
What about subranges that overlap? If someone was using something like Propeller where a function ends up split into multiple sections, then those sections are ICF'd? (I'm not 100% sure that's possible, but it seems quite plausible to me... )

**Comment by alx32** [comment]
> What about subranges that overlap? If someone was using something like Propeller where a function ends up split into multiple sections, then those sections are ICF'd? (I'm not 100% sure that's possible, but it seems quite plausible to me... )

You are correct that theoretically it's possible to have overlapping address ranges merged by something like ICF. But I think the default behavior for tooling has been to not have overlapping address ranges in dSYM's. We changed this with adding `--keep-icf-stabs` in LLD where such behavior is permitted. This change is designed to pair up with the current behavior of `--keep-icf-stabs`.

I could add the behavior to check for overlapping subranges, but I wouldn't be able to generate a test case for it as I don't think it's currently possible to generate overlapping subranges (we would need something like `--keep-icf-stabs` for Propeller). 

Any advice on how to generate a valid scenario for the overlapping subranges case or ideas how to proceed ?

**Comment by dwblaikie** [comment]
Not sure I follow the discussion about dsyms - this is a generic feature that impacts/appears in MachO as much as it does on ELF, etc.

Propeller-like debug info can be generated with `-fbasic-block-sections=all`.

There's some differences in linker behavior - lld ICF tombstones all other copies of a function/block, resolving only one copy of the references. Whereas the `gold` linker makes all references to the deduplicated code resolve to it.

So here's an example that shows functions with some parts that overlap and some that don't:
```
$ cat test.cpp
void f1() { }
void f2(bool b) {
  if (b)
    f1();
}
void f3(bool b) {
  if (!b)
    f1();
}
int main(int argc, char** argv) {
  f2(argc);
}
[hi on] blaikie@blaikie-cloudtop2:~/dev/scratch$ clang++-tot test.cpp -g -fuse-ld=gold -Wl,-icf=all -fbasic-block-sections=all && llvm-dwarfdump-tot a.out | grep -A6 DW_TAG_subprogram
...
--
0x00000037:   DW_TAG_subprogram
                DW_AT_ranges    (indexed (0x0) rangelist = 0x00000018
                   [0x0000000000000680, 0x000000000000069f)
                   [0x000000000000069f, 0x00000000000006a9)
                   [0x00000000000006a9, 0x00000000000006af))
                DW_AT_frame_base        (DW_OP_reg6 RBP)
                DW_AT_linkage_name      ("_Z2f2b")
--
0x0000004b:   DW_TAG_subprogram
                DW_AT_ranges    (indexed (0x1) rangelist = 0x00000022
                   [0x00000000000006b0, 0x00000000000006cf)
                   [0x000000000000069f, 0x00000000000006a9)
                   [0x00000000000006a9, 0x00000000000006af))
                DW_AT_frame_base        (DW_OP_reg6 RBP)
                DW_AT_linkage_name      ("_Z2f3b")
--
...
```

**Comment by alx32** [comment]
@dwblaikie - thanks for the detailed example. I can confirm that with the example provided by you we generate the same verification error.

I think we should also consider perfectly overlapping subranges as not emitting a verification warning and add a test for this scenario. Does this sound OK or did you have something else in mind ?

**Comment by dwblaikie** [comment]
Sounds good to me, thanks!

**Comment by JDevlieghere** [comment]
I ran into this myself a while ago, and I remember looking at the DWARF spec and noticing that there's no exception for function merging. I think I came to the same conclusion when I implemented this check originally. 

I agree that it makes sense to allow this , but (unless I'm wrong and don't remember or didn't look hard enough) we should raise an issue with the DWARF committee (which @dwblaikie and I are part of) to allow this in the standard.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 1]
```diff
@@ -531,14 +538,12 @@ bool DWARFVerifier::handleDebugInfo() {
   unsigned NumErrors = 0;
 
   OS << "Verifying .debug_info Unit Header Chain...\n";
-  DObj.forEachInfoSections([&](const DWARFSection &S) {
-    NumErrors += verifyUnitSection(S);
-  });
+  DObj.forEachInfoSections(
```
Comment (dwblaikie) [inline llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:541]: is this unrelated reformatting? Could you remove any instances of unrelated reformatting?

**Comment by alx32** [comment]
> we should raise an issue with the DWARF committee (which @dwblaikie, @pogo59 and I are part of) to allow this in the standard.

I'll send a proposal to [dwarf-discuss](mailto:dwarf-discuss@lists.dwarfstd.org) mailing list.

**Comment by alx32** [comment]
> I remember looking at the DWARF spec and noticing that there's no exception for function merging.

In the current [DWARF 5 spec](https://dwarfstd.org/doc/DWARF5.pdf) I can't find anything saying that overlapping address ranges are not allowed between `DW_TAG_subprogram`'s.

The only thing I can find is on page 53, line 15 where it says: 
> Bounded range entries in a range list may not overlap.

I read this as saying that if a `DW_TAG_subprogram` has `DW_AT_ranges` then the ranges of the single `DW_TAG_subprogram` cannot overlap between them. 

Am I missing something - or should the proposal be a request to clarify that overlapping address ranges are allowed between `DW_TAG_subprogram`'s ?

**Comment by dwblaikie** [comment]
> > I remember looking at the DWARF spec and noticing that there's no exception for function merging.
> 
> In the current [DWARF 5 spec](https://dwarfstd.org/doc/DWARF5.pdf) I can't find anything saying that overlapping address ranges are not allowed between `DW_TAG_subprogram`'s.
> 
> The only thing I can find is on page 53, line 15 where it says:
> 
> > Bounded range entries in a range list may not overlap.
> 
> I read this as saying that if a `DW_TAG_subprogram` has `DW_AT_ranges` then the ranges of the single `DW_TAG_subprogram` cannot overlap between them.
> 
> Am I missing something - or should the proposal be a request to clarify that identical overlapping address ranges are allowed between `DW_TAG_subprogram`'s ?

If you have overlapping subranges between subprograms, then you'll probably also end up with overlapping subranges within a single CU range list (if two functions in a single CU got ICF'd), and across CUs (if two functions across CUs got ICF'd).

It seems possible for identical overlapping subranges within a subprogram too (if two basic blocks in a single function are identical/get merged - chances are an optimizer would deduplicate those, but not guaranteed/no need to include that restriction I think - probably could come up at -O0 with basic block sections (bit of a weird feature combo, but possible))

**Comment by labath** [comment]
Just a small comment from the peanut gallery: I recently ran into some (most likely, obsolete) code in lldb which tries to handle the case where the ranges of a DW_TAG_lexical_block are not included in the ranges of its parent (block, subprogram, etc.). I was looking at the DWARF spec to see if it allows anything like this, but I couldn't find anything to support either interpretation. If you're going to be looking at strengthening the language around range relationships between various DIEs, it might be a good idea to include parent-child relationships as well. :)

**Comment by pogo59** [comment]
The parent-child aspect would be orthogonal to the OP's complaint, which is about a non-parent-child overlap, so it would want to be a separate proposal. It makes sense to me that the parent's ranges should be a superset of the child ranges, fwiw.

**Comment by alx32** [comment]
Indeed the [guidelines recommend](https://dwarfstd.org/comment.html) keeping proposals focused on a single matter:
> Each proposal should be about a single topic. Please do not combine unrelated changes in a single proposal. (Changes are unrelated if they have no interactions or dependencies.)

So looks like the parent-child aspect should be separate. 

Here is what I am thinking of proposing, feedback is appreciated:

```
BEGIN_PROPOSAL
```
# Clarification of Address Range Overlap Rules for Functions and Basic Blocks

## Background

Modern compiler optimizations and linker features can result in functions or basic blocks sharing the same address ranges in the final binary. Two common cases are:

1. Identical Code Folding (ICF) at link time, where identical functions are merged to reduce code size
2. Basic block section optimizations (e.g., `-fbasic-block-sections=all`), where basic blocks from different functions may be reordered and identical blocks merged

The current DWARF 5 specification's statement in Section 2.17.3 (p. 53, line 15) that "Bounded range entries in a range list may not overlap" is overly restrictive and doesn't reflect valid cases where identical code sequences are merged. The specification needs to differentiate between valid identical overlaps and invalid partial overlaps, both within range lists and between different `DW_TAG_subprogram` entries.

## Overview

This proposal seeks to:

1. Correct and clarify the rules regarding overlapping ranges within a single range list
2. Clarify rules for overlapping function ranges specified via DW_AT_low_pc/DW_AT_high_pc
3. Address the validity of identical overlapping ranges resulting from code merging optimizations

## Proposed Changes

In Section 2.17 (Location Descriptions), replace the text on p. 53, line 15:

From:
```
Bounded range entries in a range list may not overlap. There is no requirement that the entries be ordered in any particular way.
```

To:
```
Bounded range entries in a range list must not partially overlap. However, range entries may 
be exactly identical (having the same low and high PC values), which commonly occurs when 
compiler or linker optimizations merge identical code sequences. A partial overlap occurs 
when two ranges share some but not all addresses. There is no requirement that the entries 
be ordered in any particular way.


Similarly, when ranges are specified using DW_AT_low_pc and DW_AT_high_pc attributes, 
multiple debugging information entries (such as DW_TAG_subprogram entries) may specify 
identical ranges when they represent code that has been merged by optimizations. However, 
partial overlaps between ranges are not permitted, as this would create ambiguity in 
determining which code sequence is being executed at a given address.
```

## Dependencies

None

## References

1. LLVM Pull Request demonstrating the need for this clarification (+ additional discussion):
   https://github.com/llvm/llvm-project/pull/117952
   
   
```
END_PROPOSAL
```

**Comment by alx32** [comment]
I have sent the proposal to the dwarf-discuss mailing list [as indicated by docs](https://dwarfstd.org/comment.html#:~:text=Please%20send%20comments%20to%20the%20dwarf%2Ddiscuss%20mailing%20list).

Can we proceed with this change as of now or is the idea to wait for the outcome of the proposal ?

**Comment by dwblaikie** [comment]
One side question - this relates to both this change, and the change you folks made in the last few months related to the ICF situation with the line table? (where you wanted to make sure the line table/subprogram were used correctly together, yeah? That came from ICF and the resulting ambiguity between the .debug_info subprogram description and tnhe line table that applies - with the intent to ensure that a matched pair of line table + subprogram DIE tree were used for symbolizing, yeah?)

But lld doesn't produce these overlapping ranges, when it ICFs, it resolves only one duplicate to point to the final code, and the rest get zero'd/tombstoned, yeah?

So what situation are you in that you are getting these duplicates? Are you using gold (why that rather than lld?)? Some other linker? Is Bolt doing something esoteric? same thing as gold? 

Might be worth some design discussion about that in general.

But, yes, in the presence of gold's behavior, despite it not really being maintained so much anymore (as far as I understand it), we should probably support it...

**Comment by alx32** [comment]
> But lld doesn't produce these overlapping ranges, when it ICFs, it resolves only one duplicate to point to the final code, and the rest get zero'd/tombstoned, yeah?

This is what happens by default, yes - but the [recently added](https://github.com/llvm/llvm-project/pull/93137) `--keep-icf-stabs` does export duplicate entries in the final binary and from there `dsymutil` picks them up and generates dSYM's with overlapping ranges.

> So what situation are you in that you are getting these duplicates? 
We're only being blocked by `lld --keep-icf-stabs` + `dsymutil` => generates dSYM where `llvm-dwarfdump --verify` complains about duplicate entries. The attribute is also for this scenario.

**Comment by dwblaikie** [comment]
> > But lld doesn't produce these overlapping ranges, when it ICFs, it resolves only one duplicate to point to the final code, and the rest get zero'd/tombstoned, yeah?
> 
> This is what happens by default, yes - but the [recently added](https://github.com/llvm/llvm-project/pull/93137) `--keep-icf-stabs` lld flag does enable exporting of duplicate entries in the final binary and from there `dsymutil` picks them up and generates dSYM's with overlapping ranges.

Ah, so you're specifically seeking out this behavior, and looking to support it more.

> > So what situation are you in that you are getting these duplicates?
> 
> We're only being blocked by `lld --keep-icf-stabs` + `dsymutil` => generates dSYM where `llvm-dwarfdump --verify` complains about duplicate entries. The attribute is also only for this scenario.

Fair enough.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 2]
```diff
@@ -622,7 +627,8 @@ unsigned DWARFVerifier::verifyDieRanges(const DWARFDie &Die,
   }
 
   // Verify that children don't intersect.
-  const auto IntersectingChild = ParentRI.insert(RI);
+  bool AllowDuplicates = Die.getTag() == DW_TAG_subprogram;
```
Comment (dwblaikie) [inline llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:630]: Why restrict this to subprograms having duplicates? Presumably anything could have duplicates - if we can have subprograms with duplicates (is this duplicates within a subprogram, or between subprograms, or both?) then we can probably have lexical scopes with duplicates too? (& certainly CUs with duplicates - both within a CU and across multiple CUs)

Comment (alx32) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:630]: I was wanting to restrict this change to scenarios that I am familiar with / I'm trying to address. I am not sure about the implications for allowing duplicates for anything else. But if is clear that the right choice is to allow full overlaps for all tags - then I can make that change.

Comment (dwblaikie) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:630]: I guess a few questions:
Is this only allowing duplicates on DW_AT_ranges of DW_AT_subprograms? Or only allowing overlap /of/ subprograms /on/ CUs? I guess the latter - so that, sufficiently generalizes, handles the case of subprograms with ranges (basic block sections) that overlap with /other/ subprograms
But presumably that /doesn't/ cover the case where a subprogram's sections might overlap with itself, due to BB sections? (or perhaps it does, if there's no "self" special case - if you're allowing a subprogram to overlap with other subprograms, including itself)

But shouldn't this apply to CUs too? a CU ranges could overlap with another CU due to a subprogram in one CU overlapping with a subprogram in a different CU? (is that tested? does that somehow come out of the subprogram-only-allowed case here?)

But, equally, this should apply to lexical scopes too - should be able to make a scope in a function compiled with basic-block sections have self-duplicate ranges, or collide/duplicate with athore lexical scope's ranges in the same subprogram (by putting the same code in different places in a function, and wrapping it in lexical scopes)

Comment (alx32) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:630]: > Is this only allowing duplicates on DW_AT_ranges of DW_AT_subprograms

So this is only relating to `DW_AT_subprogram`'s - allowing them to overlap like this:
- Allows `DW_AT_ranges`'s to overlap in between different `DW_AT_subprogram`'s
- Allows identical `DW_AT_low_pc` / `DW_AT_high_pc` in between `DW_AT_subprogram`'s

> doesn't cover the case where a subprogram's sections might overlap with itself
- No, if multiple `DW_AT_ranges`'s in a `DW_AT_subprogram` overlap - this will still raise an error. **Should we cover this case also ?**

> But shouldn't this apply to CUs too? 
- `llvm-dwarfdump --error-display=details --verify` does currently not raise any errors for overlapping CU's.

> But, equally, this should apply to lexical scopes too
- Same as above, `llvm-dwarfdump --error-display=details --verify` does currently not raise any errors for overlapping lexical blocks. 

I tested using [this script](https://gist.github.com/alx32/10bde9ac21879b50a96e32e8219aa746) which generates DWARF [dumped here](https://gist.github.com/alx32/48019a09210c5e59fed32f5a124307fb).
The [output with this patch](https://gist.github.com/alx32/30ace460e5b249aca7b06891d7695e4a) shows no errors. 
The [output without this patch](https://gist.github.com/alx32/63dad3129f1b1c55265fc79fe438bcea) shows the errors for overlapping `DW_AT_subprogram`'s

Comment (dwblaikie) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:630]: Ah, yeah, looks like we don't even try to apply any verification constraints to ranges between CUs *shrug* they should probably have the same constraints, though - that exact overlap of any range is acceptable, but partial overlap is not. But ah well, I'm not pushing for improving that.

Let's look at lexical scope overlap, then... 

I think this example demonstrates almost all the overlapping cases (well, doesn't include distinct subprograms overlapping with each other, but otherwise I think it covers things):
```
int f1(bool a, bool b) {
  {
    int i = 3;
    {
      int j = 7;
      if (a)
        return 5;
    }
    {
      int j = 8;
      if (b)
        return 5;
    }
  }
  return 3;
}


int main() {
}
```
```
clang++-tot test.cpp -g -ffunction-sections -fbasic-block-sections=all -fuse-ld=gold -Wl,--icf=all
```
```
error: DIE has overlapping ranges in DW_AT_ranges attribute: [0x00000000000006a2, 0x00000000000006ae) and [0x00000000000006a2, 0x00000000000006ae)

0x0000000c:   DW_TAG_compile_unit [1] *
                DW_AT_producer [DW_FORM_strx1]  (indexed (00000000) string = "clang version 20.0.0git (git@github.com:llvm/llvm-project.git 8dd9f206b518a97132f3e2489ccc93704e638353)")
                DW_AT_language [DW_FORM_data2]  (DW_LANG_C_plus_plus_14)
                DW_AT_name [DW_FORM_strx1]      (indexed (00000001) string = "test.cpp")
                DW_AT_str_offsets_base [DW_FORM_sec_offset]     (0x00000008)
                DW_AT_stmt_list [DW_FORM_sec_offset]    (0x00000000)
                DW_AT_comp_dir [DW_FORM_strx1]  (indexed (00000002) string = "/usr/local/google/home/blaikie/dev/scratch")
                DW_AT_low_pc [DW_FORM_addr]     (0x0000000000000000)
                DW_AT_ranges [DW_FORM_rnglistx] (indexed (0x4) rangelist = 0x00000052
                   [0x0000000000000670, 0x00000000000006a2)
                   [0x00000000000006a2, 0x00000000000006ae)
                   [0x00000000000006ae, 0x00000000000006c4)
                   [0x00000000000006a2, 0x00000000000006ae)
                   [0x00000000000006c4, 0x00000000000006d0)
                   [0x00000000000006d0, 0x00000000000006d5)
                   [0x00000000000006e0, 0x00000000000006e8))
                DW_AT_addr_base [DW_FORM_sec_offset]    (0x00000008)
                DW_AT_rnglists_base [DW_FORM_sec_offset]        (0x0000000c)

error: DIE has overlapping ranges in DW_AT_ranges attribute: [0x00000000000006a2, 0x00000000000006ae) and [0x00000000000006a2, 0x00000000000006ae)

0x0000002b:   DW_TAG_subprogram [2] * (0x0000000c)
                DW_AT_ranges [DW_FORM_rnglistx] (indexed (0x0) rangelist = 0x00000020
                   [0x0000000000000670, 0x00000000000006a2)
                   [0x00000000000006a2, 0x00000000000006ae)
                   [0x00000000000006ae, 0x00000000000006c4)
                   [0x00000000000006a2, 0x00000000000006ae)
                   [0x00000000000006c4, 0x00000000000006d0)
                   [0x00000000000006d0, 0x00000000000006d5))
                DW_AT_frame_base [DW_FORM_exprloc]      (DW_OP_reg6)
                DW_AT_linkage_name [DW_FORM_strx1]      (indexed (00000003) string = "_Z2f1bb")
                DW_AT_name [DW_FORM_strx1]      (indexed (00000004) string = "f1")
                DW_AT_decl_file [DW_FORM_data1] ("/usr/local/google/home/blaikie/dev/scratch/test.cpp")
                DW_AT_decl_line [DW_FORM_data1] (1)
                DW_AT_type [DW_FORM_ref4]       (cu + 0x0087 => {0x00000087} "int")
                DW_AT_external [DW_FORM_flag_present]   (true)

error: DIE has overlapping ranges in DW_AT_ranges attribute: [0x00000000000006a2, 0x00000000000006ae) and [0x00000000000006a2, 0x00000000000006ae)

0x0000004d:   DW_TAG_lexical_block [4] * (0x0000002b)
                DW_AT_ranges [DW_FORM_rnglistx] (indexed (0x1) rangelist = 0x00000033
                   [0x0000000000000685, 0x00000000000006a2)
                   [0x00000000000006a2, 0x00000000000006ae)
                   [0x00000000000006ae, 0x00000000000006c4)
                   [0x00000000000006a2, 0x00000000000006ae))

error: DIEs have overlapping address ranges:
0x00000068: DW_TAG_lexical_block [4] * (0x0000004d)
              DW_AT_ranges [DW_FORM_rnglistx]   (indexed (0x3) rangelist = 0x0000004b
                 [0x00000000000006ae, 0x00000000000006c4)
                 [0x00000000000006a2, 0x00000000000006ae))

0x0000005a: DW_TAG_lexical_block [4] * (0x0000004d)
              DW_AT_ranges [DW_FORM_rnglistx]   (indexed (0x2) rangelist = 0x00000042
                 [0x000000000000068c, 0x00000000000006a2)
                 [0x00000000000006a2, 0x00000000000006ae))
```

Comment (alx32) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp]: Thanks for the example - the latest version addresses this.

File: llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h [hunk 3]
```diff
@@ -73,7 +73,8 @@ class DWARFVerifier {
     /// This is used for finding overlapping ranges in the DW_AT_ranges
     /// attribute of a DIE. It is also used as a set of address ranges that
     /// children address ranges must all be contained in.
-    std::optional<DWARFAddressRange> insert(const DWARFAddressRange &R);
+    std::optional<DWARFAddressRange> insert(const DWARFAddressRange &R,
+                                            bool AllowDuplicates = false);
```
[the remark this answers is not in the document]
Comment (alx32) [reply llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h]: Good point - that simplifies the change quite a bit.

**Review by dwblaikie — APPROVED** [review]

**Comment by nico** [comment]
This seems to break tests on macOS: http://45.33.8.238/macm1/97720/step_10.txt (looks like this uses `sed` features not supported by BSD sed, or at least by macOS's sed.)

Please take a look, and revert for now if it takes a while to fix.

**Comment by zeroomega** [comment]
We are seeing the test failure on 'LLVM :: tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml' in our mac-x64 builder as well.error message:

```
FAIL: LLVM :: tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml (51166 of 56656)
******************** TEST 'LLVM :: tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml' FAILED ********************
Exit Code: 2

Command Output (stderr):
--
RUN: at line 17: split-file /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml /Volumes/Work/s/w/ir/x/w/llvm_build/test/tools/llvm-dwarfdump/X86/Output/verify_no_overlap_error_icf.yaml.tmp
+ split-file /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml /Volumes/Work/s/w/ir/x/w/llvm_build/test/tools/llvm-dwarfdump/X86/Output/verify_no_overlap_error_icf.yaml.tmp
RUN: at line 18: /Volumes/Work/s/w/ir/x/w/llvm_build/bin/yaml2obj /Volumes/Work/s/w/ir/x/w/llvm_build/test/tools/llvm-dwarfdump/X86/Output/verify_no_overlap_error_icf.yaml.tmp/test.yaml | /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llvm-dwarfdump --error-display=details --verify - | /Volumes/Work/s/w/ir/x/w/llvm_build/bin/FileCheck /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/yaml2obj /Volumes/Work/s/w/ir/x/w/llvm_build/test/tools/llvm-dwarfdump/X86/Output/verify_no_overlap_error_icf.yaml.tmp/test.yaml
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llvm-dwarfdump --error-display=details --verify -
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/FileCheck /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml
RUN: at line 21: sed '0,/HighOffset: 0x6000/{s//HighOffset: 0x5999/}' /Volumes/Work/s/w/ir/x/w/llvm_build/test/tools/llvm-dwarfdump/X86/Output/verify_no_overlap_error_icf.yaml.tmp/test.yaml | /Volumes/Work/s/w/ir/x/w/llvm_build/bin/yaml2obj | /Volumes/Work/s/w/ir/x/w/llvm_build/bin/not /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llvm-dwarfdump --error-display=details --verify - | /Volumes/Work/s/w/ir/x/w/llvm_build/bin/FileCheck /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml --check-prefix=CHECK-RANGES
+ sed '0,/HighOffset: 0x6000/{s//HighOffset: 0x5999/}' /Volumes/Work/s/w/ir/x/w/llvm_build/test/tools/llvm-dwarfdump/X86/Output/verify_no_overlap_error_icf.yaml.tmp/test.yaml
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/yaml2obj
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/not /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llvm-dwarfdump --error-display=details --verify -
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/FileCheck /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml --check-prefix=CHECK-RANGES
sed: 1: "0,/HighOffset: 0x6000/{ ...": bad flag in substitute command: '}'
yaml2obj: error: unknown document type
error: -: The file was not recognized as a valid object file
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /Volumes/Work/s/w/ir/x/w/llvm_build/bin/FileCheck /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/tools/llvm-dwarfdump/X86/verify_no_overlap_error_icf.yaml --check-prefix=CHECK-RANGES

--

********************
```

Could you revert your change now and reland it with the fix please?

**Comment by dyung** [comment]
Another broken mac build bot: https://lab.llvm.org/buildbot/#/builders/23/builds/5911

**Comment by alx32** [comment]
OK, will disable the test for now.

**Comment by alx32** [comment]
Disabling via: https://github.com/llvm/llvm-project/pull/120322

**Comment by alx32** [comment]
Actual fix: https://github.com/llvm/llvm-project/pull/120330
